client/components/forms/clipboard-button/index.tsx

```diff
@@ -0,0 +1,34 @@
+import { Button, ButtonProps } from '@automattic/components';
+import classNames from 'classnames';
+import { forwardRef, ForwardRefRenderFunction } from 'react';
+
+// eslint-disable-next-line @typescript-eslint/no-empty-function
+const noop = () => {};
+
+interface ClipboardProps extends ButtonProps {
+	text: string | ( () => string );
+	onCopy?: () => void;
+}
+
+// A button which, when clicked, copies the given text to the clipboard.
+const ClipboardButton: ForwardRefRenderFunction< HTMLButtonElement, ClipboardProps > = (
+	{ className, text, onCopy = noop, ...rest },
+	ref
+) => {
+	const copyText = () => {
+		// Allow function to get text to avoid calling it on every render.
+		const copyText = typeof text === 'function' ? text() : text;
+		window.navigator.clipboard.writeText( copyText ).then( onCopy );
```

Comment:
Nit: do we need to check for the chain of objects to be defined (e.g. `window?.navigator?.clipboard?.writeText`) and/or wrap this line in a `try/catch` ? Technically `window.navigator.clipboard.writeText` is `undefined` when the page isn't loaded on HTTPS (not sure if a user could ever do that in Calypso) and the `window` object could not be defined in certain environments (e.g. certain tests, or potentially on server-side rendering)

Reply:
Good question. Given we now have unsupported browser messaging in place, the clipboard API should always be available. I'm also seeing `window.navigator` directly used in many other places in the codebase, so I think it should work okay.

But it's a pretty safe change to add as well, so I'm happy to do that if you think it's important :) The way I see it now, I'm not sure it would have an impact. And it might t be nice to use a consistent approach for accessing `window.navigator`. (If optional chaining is needed, then we should update the other use cases too)

Reply:
Since we seem to handle unsupported browsers and all tests are passing, we can leave the code as-is for now, and potentially apply a more consistent, resilient approach later